File: android/vi-api/src/main/java/lab/galaxy/yahfa/HookInfo.java
package lab.galaxy.yahfa;

import android.content.Context;

import androidx.annotation.Keep;

import com.electrolytej.vi.AbsHookRegistry;
import com.electrolytej.vi.Priority;

import java.util.ArrayList;
import java.util.List;
import java.util.ServiceLoader;

import kotlin.collections.CollectionsKt;

@Keep
public class HookInfo {
    public static List<String> hookItemNames;
    public static List<AbsHookRegistry> hookRegistry;

    static {
        hookRegistry = CollectionsKt.sortedBy(ServiceLoader.load(AbsHookRegistry.class, Thread.currentThread().getContextClassLoader()), absHookRegistry -> {
            Priority a = absHookRegistry.getClass().getAnnotation(Priority.class);
            return a != null ? a.value() : 0;
        });
        ArrayList<String> h = new ArrayList<>();
        for (AbsHookRegistry r : hookRegistry) {
            r.registerItem(h);
        }
        hookItemNames = h;
    }

    public static void attachBaseContext(Context base) {
        for (AbsHookRegistry r : hookRegistry) {
            r.attachBaseContext(base);
        }
    }
}

File: android/vi-api/src/main/java/lab/galaxy/yahfa/HookInfo.kt
@file:JvmName("HookInfo")

package lab.galaxy.yahfa

import android.content.Context
import android.util.Log
import androidx.annotation.Keep
import com.electrolytej.vi.AbsHookRegistry
import com.electrolytej.vi.Priority
import java.util.ServiceLoader

@Keep
object HookInfo {
    var hookItemNames: List<String>? = null
    var hookRegistry: List<AbsHookRegistry>? = null

    init {
        hookRegistry =
            ServiceLoader.load(
                AbsHookRegistry::class.java,
                Thread.currentThread().contextClassLoader
            )
                .sortedBy { it.javaClass.getAnnotation(Priority::class.java)?.value ?: 0 }
        try {
            val h = mutableListOf<String>()
            hookRegistry?.forEach { it.registerItem(h) }
            hookItemNames = h
        } catch (e: ClassNotFoundException) {
        } catch (e: IllegalAccessException) {
        } catch (e: InstantiationException) {
            Log.e("HookInfo", Log.getStackTraceString(e))
        }
    }

    fun attachBaseContext(base: Context) {
        hookRegistry?.forEach { it.attachBaseContext(base) }
    }
}
